File: src/structural/proxy/proxy.py

# Subject interface
class Subject:
    def request(self):
        pass

# Real Subject
class RealSubject(Subject):
    def request(self):
        print("RealSubject: Handling request.")

# Proxy
class Proxy(Subject):
    def __init__(self):
        self._real_subject = None

    def request(self):
        if self._real_subject is None:
            self._real_subject = RealSubject()
        print("Proxy: Logging request before forwarding to RealSubject.")
        self._real_subject.request()

if __name__ == "__main__":
    proxy = Proxy()
    proxy.request()

File: src/structural/proxy/Proxy.cs
namespace DesignPatterns.Structural.Proxy;

// Subject interface
public interface ISubject
{
    void Request();
}

// Real Subject
public class RealSubject : ISubject
{
    public void Request()
    {
        Console.WriteLine("RealSubject: Handling request.");
    }
}

// Proxy
public class Proxy : ISubject
{
    private RealSubject _realSubject;

    public void Request()
    {
        // Lazy initialization of RealSubject
        if (_realSubject == null)
        {
            _realSubject = new RealSubject();
        }
        
        Console.WriteLine("Proxy: Logging request before forwarding.");
        _realSubject.Request(); // Delegating to RealSubject
    }
}

// Client code
class Program
{
    static void Main()
    {
        // In a real-world application, consider using Dependency Injection (DI) to manage RealSubject.
        ISubject proxy = new Proxy();
        proxy.Request();
    }
}
